### ISSUE TITLE
Vision for pymedphys - CLIs

### ISSUE THREAD

**Issue by SimonBiggs** [pymedphys/pymedphys#15]
In GitLab by @Centrus007 on Dec 8, 2018, 11:15

This is mainly one for you, @SimonBiggs 

The PyMedPhys description states the following:

> The PyMedPhys Project aims to develop a common core package for Medical Physics in Python and foster an ecosystem of interoperable Medical Physics packages. It is inspired by our physics peers in astronomy and their [Astropy Project](http://www.astropy.org/).

Which sounds awesome to me!

With that in mind, to what extent do you envisage PyMedPhys providing (at least rudimentary) final applications? An obvious example is a DICOM anonymiser. You have mentioned before that PyMedPhys has the opportunity to export command line interfaces. Would you encourage the development of CLIs wherever they appear useful? Or would you prefer a more "library" focus?

Then there's obviously the question of GUI applications, but I figure they are out of scope.

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 8, 2018, 11:16

changed the description

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 8, 2018, 12:27

PyMedPhys is certainly a library first. Everything should bow to that. But
if it's a few one liners to turn a function into a CLI, and it's an obvious
use case (like DICOM anonymisation), then that is worthwhile. Just make
sure that the CLI isn't baked into the functions that are doing all the
work (otherwise those functions can't be used as a library). Especially if
all CLIs are namespaced under "pymedphys" then it's not too detrmintal to
just add another one.

With respect to the GUI, I'm not closed to having a "example style" GUI as
long as it is massively lite on and none of the GUI code finds its way into
the library itself.

I could imagine, for example, having a single "gui-scriptedforms" directory
which would load up a navigatable http://scriptedforms.com.au. It would be
there primarily to provide inspiration for how functions within this
library could be deployed within a centre in a user friendly way. But it
would not be the focus in any way, and by using something like
ScriptedForms the GUI code itself would be exceptionally minimal.

Cheers,
Simon

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 8, 2018, 12:38

What are your thoughts?

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 8, 2018, 13:19

That all sounds perfectly reasonable. I think that scope is probably best. But I would add:

I do think there is strong demand for open-source, user-friendly, high quality, **GUI-based** medical physics applications. Consider [QATrack+](http://qatrackplus.com/), for example. There is a nice charitable component to these apps since they can provide a dependable, useful alternative for (second & third world) departments that lack the funds/resources to purchase commercial software. Commercial options are both expensive and often not of particularly high quality. 

I fear there will be a substantial gap between a library like PyMedPhys and end use by medical physicists. Many physicists still have limited programming experience, and are often uncomfortable even with command line programs. And although I'd say in general our programming expertise seems to be improving across the board, we are *not* (yet) trained software developers...

***That is a gap that PyMedPhys itself does not need to fill.*** But, in my humble opinion, it is a gap that needs filling in order for PyMedPhys to be maximally useful.

I guess you hope/expect that this will happen organically; as a natural extension of PyMedPhys' existence (a high quality, relevant tool that is easy to leverage). I might be a little less optimistic, since there are thousands of us across the world and yet, in my estimation, rather little in the way of these sorts of projects. The gap might need a little filling from those already invested in PyMedPhys - as a different project, probably.

What do you think?

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 8, 2018, 13:22

changed the description

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 8, 2018, 13:56

You've certainly got me thinking. For simpler continuous integration
testing it would be easier to develop if it was within the same for
repository.

Personally I would prefer if the GUI was one single GUI from which
everything could be accessed.

My preference would be to use a library like the following:

https://blueprintjs.com

To create a GUI with flask and https://electronjs.org

Bundled with https://winpython.github.io so that it has its own python
installation.

...

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 8, 2018, 14:22

Also :), would you be okay if I put the requirement for at least one other programmer to review before merging to master? Do you think you would be okay with reviewing my code when I put it in? Even if it's just a glance over?

If need be, a maintainer could override it (you and I have been set to maintainers). But I am keen for us to be reviewing code before it lands in master.

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 8, 2018, 15:14

I think that would be best, yes. At the *very* least while I'm adapting to this kind of workflow.

I can certainly try!

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 8, 2018, 15:18

If the added overhead proves too restrictive, we can perhaps review this decision...

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 8, 2018, 16:47

@Centrus007 I've been thinking more and more about your idea for the GUI and I think you're right. Would you be able to have a look at the toolkits I posted in https://gitlab.com/pymedphys/pymedphys/issues/16#note_123372426 and let me know if you would be on board if we built the GUI with that?

If a GUI is being made then I would expect there to be an official `PyMedPhysGUI`. And it would sit in the same repo. And potentially be able to be started by running `pymedphys gui` in the command prompt, but also installable like any other application.

It certainly will never be as feature complete as the CLIs, and the CLIs will never be as feature complete as the library. But it will be a nice stop gap for those not programming savy.

I think every feature within the GUI should be CLI first. Meaning, everything that is worth putting in a GUI should also be worth making into a CLI. We should always make the CLI first, and make the CLI non-interactive by default. That way the GUI itself can just be a wrapper around the CLI. There should be some hint text available within the GUI that hints what command in the command prompt would achieve the same thing. ...that way the GUI can be a stepping stone to CLI, the CLI a stepping stone to using pymedphys like a python library.

I'll have a think about it, but maybe we can make the GUI creation borrow some ideas from https://github.com/chriskiehl/Gooey.

We could use electron and blueprint to make it a pretty GUI template, but what actually appears in the GUI could pull from docstrings and decorators placed around the CLI functions. I've got a few ideas. I might begin fleshing them out a little. That might actually mean, once the skeleton is in place, all we need to do is add a CLI as well as a python decorator in the code, and then hey presto, that CLI has an intuitive representation within the GUI.

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 9, 2018, 16:30

>  I've been thinking more and more about your idea for the GUI and I think you're right.

Alrighty then :smile:

>  I think every feature within the GUI should be CLI first.

Why?

>  [...] the GUI itself can just be a wrapper around the CLI.

I think I see the efficiency you're trying to bring in, especially if we agree on your second quote above. But if I'm honest, I don't yet understand the need for CLIs. Remember the end user: primarily Windows-based physicists with limited programming background. In my experience, few physicists use the command line in the clinic.

Also, are there downsides to this approach? Performance & compatibility issues, perhaps? Maybe I don't quite see the picture you're painting yet.

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 9, 2018, 16:36

> Would you be able to have a look at the toolkits I posted in https://gitlab.com/pymedphys/pymedphys/issues/16#note_123372426 and let me know if you would be on board if we built the GUI with that? 

I've never seen nor used https://blueprintjs.com or https://electronjs.org. I'm Googling and I am getting somewhere, but are there any resources you'd recommend to get up to speed on them? Their websites aren't particularly helpful (to me!). An article that discusses their use, preferably together, perhaps?

How did you come across them yourself?

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 9, 2018, 17:09

A CLI is something that can be easily run from the command prompt, and is accessible from most programming languages. In Linux world they can be piped together in bash which is also massively convenient.

Essentially, when one is proficient in the command line, tasks done via CLI end up all happening massively quicker.

A command line first approach means that any language that can run system calls (most) can call the CLI and programmatically interface in a clean way with the library. Making it easy for end users to string together tasks with a few commands, maybe create their own bat files to run with windows task scheduler etc.

By default the CLI will require far less code to maintain and it will be easier to test with automated testing. By making the GUI inherit from the CLI and build on top of the CLI first, then we test the CLI, test that the GUI calls the CLI templates accurately, instead of needing to create loads of tests that control a mouse and keyboard or something similar (like selenium) and click all over the place.

Electron js was created by GitHub to build Atom. GitHub was purchased by Microsoft, and Microsoft used electron js to build VS Code. Therefore electron js is now maintained by Microsoft. Another tool built with electron js is slack.

Blueprint JS is a components library which makes all the pretty interaction widgets. They're designed for slick interfaces on the desktop. When googling "best react component library" it appears near the top, and is the only one of the top the is "desktop user first", it is designed for fitting a lot of information in a small space. It was recommended by one of the coders within the pyodide crew.

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 9, 2018, 17:12

Biggest benefit of using tools like electronjs and blueprint JS is the end result looks very slick and professional without having to focus on visual design.

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 9, 2018, 17:34

The biggest benefit for CLIs really is just how fast and efficient they are to use.

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 9, 2018, 17:43

Okay, combining all of that with my own reading:

For now at least, **I'm convinced on all counts**. I say "for now", since it'll be contingent on further information...

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 9, 2018, 17:57

I guess the last point about the target audience. Although the largest audience would be non-technical users, the specific audience that would have the capability to make PyMedPhys itself better, grow PyMedPhys, and make it better for all users are developers. So, even though they make up a small portion of the total audience, it is that group that, for the health of PyMedPhys and its sustainability long term, needs to have the first and foremost focus.

It is that group that appreciates CLIs, REST APIs, and programming libraries. So for the long term quality and maintainability of PyMedPhys those sorts of tools need to be prioritised. If it can be made that a GUI builds off of a CLI then we can focus on the CLI and get the GUI bundled in.

... we'll see. ...and we'll see how that further information goes :).

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 9, 2018, 18:00

(the number of edits on that last comment of mine. Gee... I should stop replying on my phone)

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 9, 2018, 18:18

Fair enough. :)

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 9, 2018, 18:34

By the way, an example of one of the components exported by blueprintjs is it's spreadsheet like table element:

https://blueprintjs.com/docs/#table/features

(Designed for viewing on the desktop, not mobile)

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 9, 2018, 18:45

@Centrus007 also another thing for GUIs backed by CLIs. To provide support online it is far easier to say, "can you copy the following into a command prompt and tell me what it says" (which I quickly wrote on my phone) as opposed to, can you follow these 7 screenshots (that took me 5 minutes to put together on my desktop) and then take a screenshot of the following two screens and send them to me.

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Jan 5, 2019, 06:19

closed
